### ISSUE TITLE
Clarify API and flow for Partials

### ISSUE THREAD

**Issue by tomalec** [Starcounter/Starcounter#3538]
As discussed on today's meeting with @chrhol, @per-samuelsson, @warpech, @miyconst and me, we agreed to take some steps towards simplifying/cleaning up the flow and code for partials.

We well provide more self-explanatory methods, to work with partials and mapping:
 - `Partial.Get` - like `Self.GET` which also handles objects as parameters - function used to receive partial's result
 - `Partial.Register` - `Handle.GET` for Partials
 - `Partial.Map` - `UriMapping.OntologyMap` + `UriMapping.Map`

Then:
 - user will be aware what such methods are used for,
 - we will use term `partial` consistently across client and server
 - even if there is technical similarity between Partial and regular response, we will clearly distinguish one from another,
 - JSON merger will simply merge only responses for `Partial.Get`, but not all (external) responses,
 - we could use objects as parameters without a need to serialize it to URI string,
 - mapping takes place for partial rather than full trees (JSON/HTML docs),
 - this API could be used existing one,
 - it's fully backward compatible

The plan is to implement and use it in our code, to check how it feels before we start documenting, advertising, and shifting our clients to it.

**Comment by per-samuelsson** [comment]
@tomalec 

I'm still not 100% about this API, but as I said, it's at least better then the one we have now, and as you wrote, let's try it out and see what it feels like.

What I really wanted to add here was a big :+1: for the way you described it and the great listing of motivations and aims. Really put the finger on what we are trying to improve!

**Comment by warpech** [comment]
@aloscha, could you please make this? 

As proof-of-concept, create a class `Partial.cs` in Barcodes app with three static methods (explained by @tomalec above). These methods should just be one-liners that wrap the existing functionality used in showing partials from other apps.

When done' verify that it works with other apps (test Launcher + Products + Barcodes).

After we are happy with this proof-of-concept, these static methods can be added tothe existing [`Partial.cs`](https://github.com/Starcounter/Starcounter/blob/develop/src/Starcounter/Partial.cs) class in Starcounter.

**Comment by aloscha** [comment]
@warpech I create proof-of-concept on Image application on branch - https://github.com/StarcounterSamples/Images/tree/Partials_Flow_%233538

I tested that with Chatter on Master branch and it works

**Comment by aloscha** [comment]
And tested with Products app (Master branch)

**Comment by warpech** [comment]
@per-samuelsson, @miyconst, could you please review the implementation linked above?

I am happy with the job done by @aloscha, but still not totally happy with the design. The existence of `Partial.Get`, `Partial.Register` and `Partial.Map` in the app still looks weird. It is certainly less weird that dozens of `Handle.GET`, `Self.GET` and `UriMapping.OntologyMap`, but does not feel to be the sweet spot.

**Comment by miyconst** [comment]
Hm, as I remember we agreed to:

- Use `Handle.GET` or `Handle.POST` for partials which are exposed to client and are **not** available for mapping.
- Use `Partial.Register` for partials which are for internal usage only, those are available for mapping.

In the current implementation I don't see any `Handle.GET`. How do I know which are the application entry points and which are the partials?

Also there are two methods: `Register` and `RegisterPartials` both use `Partial.Register`, does not make much sense for me.

**Comment by warpech** [comment]
This is true. Handlers in lines https://github.com/StarcounterSamples/Images/blob/Partials_Flow_%233538/src/Images/Api/MainHandlers.cs#L29-L53 should be created using `Handle.GET`, not `Partial.Register`. @aloscha can you fix that?

`RegisterPartials` is just an artifact from the current `master` branch. It is not part of the work done in this issue.

**Comment by aloscha** [comment]
Fixed - https://github.com/StarcounterSamples/Images/commit/ce76dbc72efb3229e5805c839f549255e0f30c61

**Comment by warpech** [comment]
I propose that `Partial.Register` should not allow providing `HandlerOptions`. Rather than that, it should have hardcoded `new HandlerOptions() { SelfOnly = true }`.

It also should not allow specifying method. Partials are always `"GET"`.

**Comment by aloscha** [comment]
Maybe instead of null as a default value of HandlerOptions use what you say @warpech `new HandlerOptions() { SelfOnly = true }`?

**Comment by warpech** [comment]
Fine for me

**Comment by miyconst** [comment]
@aloscha you are right, `Partials.Register` should never register a handler which is accessible from outside.

**Comment by miyconst** [comment]
Also consider simplifying mapping, so it will look like:

```
Partial.Map<Simplified.Ring3.Person>("/images/concept/{?}");
```

**Comment by aloscha** [comment]
https://github.com/StarcounterSamples/Images/commit/06b381f3ac5483efcf1a1096caf17405ba0ea73b

@miyconst We could consider in next step, I think now we need to create just proof-of-concept (simple wrapper)

**Comment by warpech** [comment]
> @miyconst We could consider in next step, I think now we need to create just proof-of-concept (simple wrapper)

This is something already supported in OntologyMap, so I think that `Partial.Map` can wrap it as well: https://github.com/Starcounter/Starcounter/blob/c7f7424f7fb3984d13f5dbbe0184611ab4b843f7/src/Starcounter/Apps/OntologyAndOrdinaryMapping.cs#L464

**Comment by aloscha** [comment]
Indeed @warpech have good eye for that

**Comment by aloscha** [comment]
Latest version is on branch - https://github.com/StarcounterSamples/Images/tree/Partials_Flow_%233538

**Comment by chrhol** [comment]
I've been working on #3680, and haven't gotten very far. I got stuck on the jsonmerger, so improving the partial flow is getting a bit more important I think. 

 I will spend some time trying out the implementation here (I should have done that a long time ago...)

**Comment by chrhol** [comment]
I pointed @Starcounter-Jack to this issue today, and I (tried to :smile:) explain what problems we have and what we were experimenting with. 

His suggestion was to start from the other direction, with documenting how blending and mapping should work, together with examples on what happens on different levels (client, gui, server, orchestrator), and what the requirements are instead of starting with changing the syntax. 

He will make some examples and drawings, based on the gui and viewmodels from the video (https://www.youtube.com/watch?v=pcHmUa0GO0Q), which we will add (I guess, at least initially) on rebelslounge gitbook.

**Comment by warpech** [comment]
I think this is a good approach. Just please check this with @Starcounter-Jack every few days or so, so we are not stuck on this, because the developer UX on partials really needs improvement.

**Comment by chrhol** [comment]
Yeah, will do that. I think the part about the **Orchestrator** is the most important, since this is something that is not discussed that much, mostly heard about in bits and pieces around the office.
